### DISCUSSION TITLE
How to duplicate ‘/Root’, except ‘/Metadata’, from PdfReader to PdfWriter.

### DISCUSSION THREAD

**Discussion by macdeport** [py-pdf/pypdf#2691] [Q&A]
I have to clone `PdfReader` `viewer_preferences` in `PdfWriter`: what is the best way to do?

More generally I try to duplicate ‘/Root’, except ‘/Metadata’, from `PdfReader` to `PdfWriter`.

**Comment by stefan6419846** [comment]
Please do not open duplicates - you already asked the same in #2688.

The classes provide access to `/Root` as `root_object` (example: `PdfWriter(clone_from="file.pdf").root_object`, which basically is just a dictionary: https://pypdf.readthedocs.io/en/stable/modules/PdfWriter.html#pypdf.PdfWriter.root_object If there is a metadata-related key, you can do whatever you want with it if you are aware what you are doing.

**Reply by pubpub-zz** [reply]
"/Root" being the top object is duplicated through init with "clone_from" argument
once you have duplicated data you can use `._replace_object( **indirect_reference**, pypdf.generic.NullObject())` to delete the object

**Reply by macdeport** [reply]
> Please do not open duplicates - you already asked the same in #2688.

Sorry , I closed the discussion prematurely and was unable to reopen it..

https://github.com/py-pdf/pypdf/discussions/2691#discussioncomment-9583654

I didn't manage to handle an indirect object properly

```Python
'''
root={'/Type': '/Catalog', '/Pages': IndirectObject(2, 0, 4366473040), '/Outlines': IndirectObject(43, 0, 4366473040), 
'/PageMode': '/UseOutlines', '/OutputIntents': [IndirectObject(48, 0, 4366473040)],
'/Metadata': IndirectObject(50, 0, 4366473040), '/Lang': 'fr', '/ViewerPreferences': {'/DisplayDocTitle': True}, 
'/MarkInfo': {'/Marked': True}, '/StructTreeRoot': IndirectObject(51, 0, 4366473040)}
'''

writer = PdfWriter(clone_from = fp)
root = writer.root_object
writer._replace_object(root['/Metadata'], pypdf.generic.NullObject())

Traceback (most recent call last):
  File "/var/folders/ps/z7flxvdj3b97p9_07lknl6dc0000gn/T/com.barebones.bbedit/BBEditRunTemp-untitled text 8.py", line 172, in <module>
    writer._replace_object(root['/Metadata'], pypdf.generic.NullObject())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pypdf/_writer.py", line 367, in _replace_object
    gen = self._objects[indirect_reference - 1].indirect_reference.generation  # type: ignore
                        ~~~~~~~~~~~~~~~~~~~^~~
TypeError: unsupported operand type(s) for -: 'DecodedStreamObject' and 'int'

```

**Reply by pubpub-zz** [reply]
can you please provide your test pdf.

**Reply by macdeport** [reply]
It's a ordinary private PDF/A document produced by an institution that you may already have yourself: an recent EDF bill (download version).

If this is not the case, we will consider another solution.

**Reply by macdeport** [reply]
De façon surprenante ceci fonctionne :

```Python
from pypdf import PdfReader, PdfWriter
import pypdf

fp = 'input.pdf'; fpo = 'output.pdf'
writer = PdfWriter(clone_from=fp)
root = writer.root_object;print(f'{root=}\n')
'''
root={'/Type': '/Catalog', '/Pages': IndirectObject(2, 0, 4366473040), 
      '/Outlines': IndirectObject(43, 0, 4366473040), '/PageMode': '/UseOutlines',
      '/OutputIntents': [IndirectObject(48, 0, 4366473040)], 
      '/Metadata': IndirectObject(50, 0, 4366473040), '/Lang': 'fr', 
      '/ViewerPreferences': {'/DisplayDocTitle': True}, '/MarkInfo': {'/Marked': True}, 
      '/StructTreeRoot': IndirectObject(51, 0, 4366473040)}
'''

if md := root.get('/Metadata'):
	if isinstance(md, pypdf.generic._base.IndirectObject):
		print(f"{root['/Metadata'].get_object()=}")
		print(f"{md.get_object()=}\n")
		del root['/Metadata']
		#writer._replace_object(root['/Metadata'], pypdf.generic.NullObject())
		writer._replace_object(md, pypdf.generic.NullObject())
		print(f"{root.get('/Metadata')=}\n")
		print(f'{writer.metadata=}\n')
		print(f'{root=}\n')

with open(fpo, 'wb') as f:
	writer.write(f)

reader = PdfReader(fpo);root = reader.root_object
print(f'{root=}\n')

```
`writer._replace_object(md, pypdf.generic.NullObject()) ` works, while `writer._replace_object(root['/Metadata'], pypdf.generic.NullObject())` is rejected

As it is paradoxical a logical explication will be found...

Despite deletion the object '/Metadata' continue to point to the same indirect object (???)

Output of the previous code:
```Python
root={'/Type': '/Catalog', '/Pages': IndirectObject(2, 0, 4346905552), 
      '/Outlines': IndirectObject(34, 0, 4346905552), '/PageMode': '/UseOutlines', 
      '/OutputIntents': [IndirectObject(38, 0, 4346905552)], '/Metadata': IndirectObject(40, 0, 4346905552), 
      '/Lang': 'fr', '/ViewerPreferences': {'/DisplayDocTitle': True}, '/MarkInfo': {'/Marked': True}, 
      '/StructTreeRoot': IndirectObject(41, 0, 4346905552)}

root['/Metadata'].get_object()={'/Type': '/Metadata', '/Subtype': '/XML'}
md.get_object()={'/Type': '/Metadata', '/Subtype': '/XML'}

root.get('/Metadata')=None

writer.metadata={'/Creator': 'OpenText Exstream Version 16.4.0 64-bit', '/CreationDate': 'D:20240523002508Z', 
                 '/Author': 'Registered to: EDF     ', '/Title': 'FACTURE_SIMM_NOUVEAU_FACO'}

root={'/Type': '/Catalog', '/Pages': IndirectObject(2, 0, 4346905552), 
      '/Outlines': IndirectObject(34, 0, 4346905552), '/PageMode': '/UseOutlines', 
      '/OutputIntents': [IndirectObject(38, 0, 4346905552)], '/Lang': 'fr', 
      '/ViewerPreferences': {'/DisplayDocTitle': True}, '/MarkInfo': {'/Marked': True}, 
      '/StructTreeRoot': IndirectObject(41, 0, 4346905552)}

```

https://github.com/py-pdf/pypdf/discussions/2691#discussioncomment-9583654

How to access "/Root" duplicate?

**Reply by pubpub-zz** [reply]
> It's a ordinary private PDF/A document produced by an institution that you may already have yourself: an recent EDF bill (download version).
> 
> If this is not the case, we will consider another solution.

Sorry I do not have such pdf. You should be able to generate an empty one using `.remove_text()`

**Reply by pubpub-zz** [reply]
> `writer._replace_object(md, pypdf.generic.NullObject()) ` works, while `writer._replace_object(root['/Metadata'], pypdf.generic.NullObject())` is rejected

root.get('/Metadata') provides the real data stored in the dictionnary : here it is a IndirectObject
for convenience, root["/Metadata"] calls also `.get_object()` : inhere it will returned the DictionnaryObject pointed behind the IndirectObject. 
if DictionnaryObject was stored directly in root_object, `.get_object()` will returned it.

using _replace_object you will replace the object "pointed". If you were just doing `del root["/MetaData"]` the DictionnaryObject in the `._objects` array  would remain and will be written in the pdf file when calling `.write()`. if you prefer you can do both

**Comment by macdeport** [comment]
> Sorry I do not have such pdf. You should be able to generate an empty one using .remove_text()

Below this file:
[anonyme0.pdf](https://github.com/py-pdf/pypdf/files/15489785/anonyme0.pdf)
